internal/api/core/core_test.go

```diff
@@ -0,0 +1,126 @@
+	return apiConfig.Config{
+		Security: apiConfig.Security{
+			Readonly:      false,
+			EnableAuth:    false,
+			EncryptionKey: secret.Hidden(hex.EncodeToString([]byte("=tW$56zytgB&3jN2E%7-+qrGZE?v6LCc"))),
+		},
+		Plugin: apiConfig.Plugin{
+			Path:        filepath.Join(projectPath, "plugins"),
+			ArchivePath: filepath.Join(projectPath, "plugins-archive"),
+		},
+		Database: apiConfig.Database{
+			File: &apiConfig.File{
+				Folder:        "./test",
+				Extension:     apiConfig.JSONExtension,
+				CaseSensitive: true,
+			},
+		},
+	}
+}
+
+func withServer(t *testing.T, config apiConfig.Config, testFunc func(*httpexpect.Expect)) {
+	runner, persistenceManager, err := New(config, false, prometheus.NewRegistry(), "")
+	assert.NoError(t, err)
+
+	handler, err := runner.HTTPServerBuilder().BuildHandler()
+	assert.NoError(t, err)
+
+	server := httptest.NewServer(handler)
+
+	defer persistenceManager.GetPersesDAO().Close()
+	defer server.Close()
+
+	expect := httpexpect.WithConfig(httpexpect.Config{
+		BaseURL:  server.URL,
+		Reporter: httpexpect.NewAssertReporter(t),
+	})
+	testFunc(expect)
+}
+
+func sendPreflightRequest(expect *httpexpect.Expect, origin string) *httpexpect.Response {
+	return expect.OPTIONS(fmt.Sprintf("%s/projects", utils.APIV1Prefix)).
+		WithHeader("Origin", origin).
+		Expect()
+}
+
+func sendRequest(expect *httpexpect.Expect, origin string) *httpexpect.Response {
+	return expect.GET(fmt.Sprintf("%s/projects", utils.APIV1Prefix)).
+		WithHeader("Origin", origin).
+		Expect()
+}
+
+func TestCORSDefaultConfig(t *testing.T) {
+	config := defaultConfig()
+	config.Security.CORS = apiConfig.CORSConfig{
+		Enable: true,
+	}
+	withServer(t, config, func(expect *httpexpect.Expect) {
+		resp := sendPreflightRequest(expect, "https://github.com")
+		resp.Status(http.StatusNoContent)
+		resp.Header("Access-Control-Allow-Origin").IsEqual("*")
+		resp.Header("Access-Control-Allow-Methods").IsEqual("OPTIONS, GET, POST")
+
+		resp = sendRequest(expect, "https://github.com")
+		resp.Status(http.StatusOK)
+		resp.Header("Access-Control-Allow-Origin").IsEqual("*")
+	})
+}
+
+func TestCORSCustomConfig(t *testing.T) {
+	config := defaultConfig()
+	config.Security.CORS = apiConfig.CORSConfig{
+		Enable:       true,
+		AllowOrigins: []string{"https://gitlab.com", "https://github.com"},
+		AllowMethods: []string{"OPTIONS", "GET"},
+	}
+	withServer(t, config, func(expect *httpexpect.Expect) {
+		resp := sendPreflightRequest(expect, "https://github.com")
+		resp.Status(http.StatusNoContent)
+		resp.Header("Access-Control-Allow-Origin").IsEqual("https://github.com")
+		resp.Header("Access-Control-Allow-Methods").IsEqual("OPTIONS,GET")
+
+		resp = sendRequest(expect, "https://github.com")
+		resp.Status(http.StatusOK)
+		resp.Header("Access-Control-Allow-Origin").IsEqual("https://github.com")
+
+		resp = sendRequest(expect, "https://google.com")
+		resp.Status(http.StatusOK)
+		resp.Header("Access-Control-Allow-Origin").IsEmpty()
+	})
+}
```

Comment:
Maybe we are missing a test in which the preflight request fails, WDYT?

Reply:
I added an addition test for a preflight request with an invalid origin, but I'm not a CORS expert to be certain of what exactly is expected (the CORS middleware should be right about it though!). My understanding is that the request itself succeeds (i.e., it does return a 2xx code), that is the browser that is ultimately failing it if some headers are missing/incorrect.

Reply:
Yes would be hard to make a failing cors request actually as it's something forbidden only in a browser. Maybe creating a selenium test and I'm not even sure that localhost wouldn't be just whitelisted. Ok to not have failing tests here as impossible to me.

Reply:
We can do a [preflight request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/CORS#preflighted_requests) that uses the OPTIONS method and assert that it returns a 204 or not.

Reply:
@jgbernalp you mean make a preflight request when CORS are disabled ?

Reply:
I meant to test that having CORS disabled does not change the current behavior. But I think the happy path would be sufficient as with CORS disabled the middleware is not added.